lib/semantics/semantics.cc

```diff
@@ -168,6 +298,24 @@ Scope &SemanticsContext::FindScope(parser::CharBlock source) {
   }
 }
 
+void SemanticsContext::PushConstruct(const ConstructNode &construct) {
+  constructStack_.emplace_back(construct);
+}
+
+void SemanticsContext::PopConstruct() {
+  CHECK(!constructStack_.empty());
+  constructStack_.pop_back();
+}
+
+bool SemanticsContext::InsideDoConstruct() const {
+  for (const ConstructNode construct : constructStack_) {
```

Comment:
Rather than looping over the entries in the stack and looking for `DO` loops, you could maintain a count of active `DO` loops.

Reply:
Actually, that was my original implementation.  When I reviewed that with Tim, he suggested using a stack.  The stack is heavier weight, but it has the promise of being useful in other kinds of semantic analysis.

Reply:
You can still maintain a counter.

Reply:
Or maybe a cumulative ancestor set for all ancestor types, not just do loops.  That is, for top level constructs, this set is empty.  For nested constructs, the set is the parent's set plus the parent's type.  That way, queries of the form "this construct has at least one kind X ancestor" can be answered without searching.  And queries of the form "this construct is immediately nested in a kind X parent' can also be directly answered by looking one up the stack.  Those two queries probably cover a large majority of all potential queries.

Reply:
Val suggestion is more efficient compared to the stack approach. However one would still have to traverse up get a count of active `DO` loops. So a cumulative ancestor set + depth record for `DO` (and any other construct that needs its depth saved) seems like the best idea.

Reply:
Thanks for the suggestions.  One of the checks I plan to implement (although not in this pull request) is to semantic checking for ```CYCLE do-construct-name```.  For that case, I'll need to walk up the currently active set of nested ```DO``` constructs looking for one with ```do-construct-name```.  For just checking a lone ```CYCLE``` statement, a counter is sufficient and more efficient.  Since I don't think I can get rid of the stack, and since I believe that it's sufficient for the cases of ```CYCLE``` statements with and without ```do-construct-name```s, I plan to just go with a stack.  I'll keep the notion of counters in mind as I add the rest of the checks, though.